shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java

```diff
@@ -186,4 +186,124 @@ public int getOffsetBefore(CharSequence text, int offset) {
 
     return offset - deleteCharCount;
   }
+
+  /**
+   * Gets the offset of the next character following the given offset, with consideration for
+   * multi-byte characters.
+   */
+  public int getOffsetAfter(CharSequence text, int offset) {
+    final int len = text.length();
+
+    if (offset >= len - 1) {
+      return len;
+    }
+
+    int codePoint = Character.codePointAt(text, offset);
+    int nextCharCount = Character.charCount(codePoint);
+    int nextOffset = offset + nextCharCount;
+
+    if (nextOffset == 0) {
+      return 0;
+    }
+
+    // Line Feed
+    if (codePoint == LINE_FEED) {
+      codePoint = Character.codePointAt(text, nextOffset);
+      if (codePoint == CARRIAGE_RETURN) {
+        ++nextCharCount;
+      }
+      return offset + nextCharCount;
+    }
+
+    // Flags
+    if (isRegionalIndicatorSymbol(codePoint)) {
+      codePoint = Character.codePointAt(text, nextOffset);
+      nextOffset += Character.charCount(codePoint);
+      int regionalIndicatorSymbolCount = 0;
+      while (nextOffset < len && isRegionalIndicatorSymbol(codePoint)) {
+        codePoint = Character.codePointAt(text, nextOffset);
+        nextOffset += Character.charCount(codePoint);
+        regionalIndicatorSymbolCount++;
+      }
+      if (regionalIndicatorSymbolCount > 0 && regionalIndicatorSymbolCount % 2 == 0) {
```

Comment:
When moving the caret to right, there is no difference for odd and even regional indicator symbols. except that:
1. If there is only one, delete it.
2. if the previous character is also a regional indicator symbol, move the caret only one position.

It is because for example "I R R" was rendered like "[ I R ] R" on the screen.
It is working in the tests correctly because you are counting the number of symbols after the current symbol. But the logic is incorrect

Reply:
Thanks for catching this and for the clear explanation!  I'll fix this and post back.

Reply:
I fixed this in https://github.com/flutter/engine/pull/17420/commits/48b0de9a53b65c263174012af6c732051b4251bb.  I think I understand how these region indicators pair up now, before I wasn't sure exactly what the while loop was for.

I also added a separate test to check this in detail.  Let me know if anything still doesn't look right.